### ISSUE TITLE
Add compiler trick to avoid allocation of frozen String literals

### ISSUE THREAD

**Issue by donv** [jruby/jruby#1862]
Hi!

I just saw this presentation which mentions the Ruby 2.1 trick to re-use immutable String objects for frozen String literals:

https://www.youtube.com/watch?feature=player_detailpage&v=d2QdITRRMHg#t=3570

It is heavily used in Rails and JRuby should take advantage of it.

Example code:

With Ruby 2.1:

```ruby
2.1.2 :002 > 5.times{p 'a'.object_id}
70119468854480
70119468854360
70119468854280
70119468854200
70119468854120
 => 5 
2.1.2 :003 > 5.times{p 'a'.freeze.object_id}
70119476885680
70119476885680
70119476885680
70119476885680
70119476885680
 => 5 
```

With JRuby 1.7.13:

```ruby
jruby-1.7.13 :001 > 5.times{p 'a'.freeze.object_id}
2042
2044
2046
2048
2050
 => 5 
```

With JRuby master:

```ruby
jruby-head :001 > 5.times{p 'a'.freeze.object_id}
2048
2050
2052
2054
2056
 => 5 
```

**Comment by nehresma** [comment]
This is still open but it appears that it was implemented at some point.  I haven't gone through the changelog to identify which release, but:

```
jruby-9.1.1.0 :001 > 5.times{p 'a'.freeze.object_id}
2048
2048
2048
2048
2048
 => 5 
jruby-9.1.1.0 :002 > 5.times{p 'a'.object_id}
2050
2052
2054
2056
2058
```

Possibly this could be closed?

**Comment by headius** [comment]
Good catch, thanks @nehresma! I am going to mark this as fixed in 9.1, because before that it was only partially implemented.
